### PR TITLE
Fix Satellite Assembly loading

### LINKED ISSUE #20979: LoadFrom Resource Fallback search fails to search next to current executing assembly
For .NET Core the primary location to look for satellite resource assemblies should be in the culture directory next to the current executing assembly.

This directory is not currently being searched unless it happens to be already on the search path

@jkotas @jeffschwMSFT @tarekgh @vitek-karas FYI

See dotnet/core#2041

### LINKED ISSUE #24211: Satellite Lookup on case sensitive file systems...
@tarekgh Made a request for searching for Satellite Assemblies using case insensitive path search.

Initial prototype of the case insensitive search illustrated the complexity of such an approach.

The prototype case insensitive search code was removed and this issue was opened for the design discussion.

@jkotas made the alternative suggestion that we control the build tooling.  We could simply `lower case` the culture directories....

See discussion which started here https://github.com/dotnet/coreclr/pull/24191#discussion_r277793894

### PR DESCRIPTION
**Pull request by sdmaclea**
When loading satellite assemblies, we should probe next to the parent
assembly and load into the same AssemblyLoadContext as the parent
assembly.

Disable fallback probing for satellite assemblies.

Add AssemblyLoadContext.Resolving handler to probe for satellite
assemblies next to parent

Fixes #20979
Fixes #24211

### CODE REVIEW

**Comment by sdmaclea** [comment]
I have not yet tested this. I will write tests in corefx, and manually test before merging this.

* I wanted to make sure we were OK w/ writing this in managed code.
* The `FindCaseInsensitiveSatellitePath()` code could be simpler if a *.GetDirectories() existed in CoreCLR. It seems these are implemented in CoreFX on top of the same InterOp.

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+                                string assemblyPath = $"{parentDirectory}/{entryNameString}/{assembly}.dll";
+
+                                if (Internal.IO.File.InternalExists(assemblyPath))
+                                    return assemblyPath;
+                            }
+                        }
+                        finally
+                        {
+                            if (dirHandle != IntPtr.Zero)
```
Comment (tarekgh) [inline src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: >if (dirHandle != IntPtr.Zero) [](start = 28, length = 29)

Is it possible dirHandle be equal IntPtr.Zero here?

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+{
+    public partial class AssemblyLoadContext
+    {
+        // Find satellite path using case insensitive culture name
+        internal static unsafe string? FindCaseInsensitiveSatellitePath(string parentDirectory, string cultureName, string assembly)
+        {
+            int bufferSize = Interop.Sys.GetReadDirRBufferSize();
+            byte[]? dirBuffer = null;
+            try
+            {
```
Comment (tarekgh) [inline src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: why don't move this try after ArrayPool<byte>.Shared.Rent(bufferSize); and then avoid checking for null in the finally block?

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                    {
+                        try
+                        {
+                            Interop.Sys.DirectoryEntry dirent;
+                            while (Interop.Sys.ReadDirR(dirHandle, dirBufferPtr, bufferSize, out dirent) == 0)
+                            {
+                                if (dirent.InodeType != Interop.Sys.NodeType.DT_DIR)
+                                    continue;
+
+                                Span<char> nameBuffer = stackalloc char[Interop.Sys.DirectoryEntry.NameBufferSize];
```
Comment (tarekgh) [inline src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: >Span<char> nameBuffer = stackalloc char[Interop.Sys.DirectoryEntry.NameBufferSize]; [](start = 32, length = 83)

would make sense move this stackalloc outside the while loop?

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                                ReadOnlySpan<char> entryName = dirent.GetName(nameBuffer);
+
+                                if (cultureName.Length != entryName.Length)
+                                    continue;
+
+                                string entryNameString = entryName.ToString();
```
Comment (tarekgh) [inline src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: >string entryNameString = entryName.ToString(); [](start = 32, length = 46)

I think there is a way to avoid this allocation. maybe you can get the Span of cultureName and then compare the spans instead of comparing strings.

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+                                if (!cultureName.Equals(entryNameString, StringComparison.InvariantCultureIgnoreCase))
+                                    continue;
+
+                                string assemblyPath = $"{parentDirectory}/{entryNameString}/{assembly}.dll";
```
Comment (tarekgh) [inline src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: >string assemblyPath = $"{parentDirectory}/{entryNameString}/{assembly}.dll"; [](start = 32, length = 76)

would we use span for this one too? or we have to allocate anyway?

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+            if (parentAlc == null)
+                return null;
+
+            string parentDirectory = Path.GetDirectoryName(parentAssembly.Location)!;
+
+            string assemblyPath = $"{parentDirectory}/{cultureName}/{assemblyName.Name}.dll";
+            if (Internal.IO.File.InternalExists(assemblyPath))
+            {
+                Assembly satelliteAssembly = parentAlc.LoadFromAssemblyPath(assemblyPath);
```
Comment (tarekgh) [inline src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: >Assembly satelliteAssembly [](start = 16, length = 26)

nit: Maybe you can define 

Assembly satelliteAssembly = null; 

before the if statement, then don't check for null values inside the if at all and let the method return satelliteAssembly.

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+                                Span<char> nameBuffer = stackalloc char[Interop.Sys.DirectoryEntry.NameBufferSize];
+                                ReadOnlySpan<char> entryName = dirent.GetName(nameBuffer);
+
+                                if (cultureName.Length != entryName.Length)
+                                    continue;
+
+                                string entryNameString = entryName.ToString();
+
+                                if (!cultureName.Equals(entryNameString, StringComparison.InvariantCultureIgnoreCase))
```
Comment (tarekgh) [inline src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: >InvariantCultureIgnoreCase [](start = 90, length = 26)

should we use OrdinalIgnoreCase instead? using Invariant can have some security implication I guess.

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+using System.Buffers;
+using System.IO;
+using System.Reflection;
+
+namespace System.Runtime.Loader
+{
+    public partial class AssemblyLoadContext
+    {
+        // Find satellite path using case insensitive culture name
+        internal static unsafe string? FindCaseInsensitiveSatellitePath(string parentDirectory, string cultureName, string assembly)
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: I am not sure whether we need to be do the case-insensitive probing here. I know we have it in some places, but it is expensive, not done consistently and a pit of failure.

Can we just document that the case has to match?

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs [hunk 9]
```diff
@@ -100,6 +100,8 @@ private protected AssemblyLoadContext(bool representsTPALoadContext, bool isColl
                 _id = s_nextId++;
                 s_allContexts.Add(_id, new WeakReference<AssemblyLoadContext>(this, true));
             }
+
+            Resolving += ResolveSatelliteAssembly;
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: It does not sound right that we are always unconditionally subscribing to event. Should this be a callback from the runtime instead?

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs (hunk 8 above)
Comment (tarekgh) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: Actually have strict case naming causing some problems. Sometime the globalization name can be returned in different casing. so it will be hard for people to always have exact casing. one example is zh-Hans/Hant which can be also valid as zh-HANS and zh-HANT. of course we can strict casing in the globalization layer but we'll have to add a lot of special cases as we encounter them. looks we only probe for other casing only if we couldn't load the satellite assembly which I think can be acceptable perf wise.

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: > Sometime the globalization name can be returned in different casing. 

Is there a default/standard casing for the culture name IDs? Can we normalize to that in the tooling? 

> Looks we only probe for other casing only if we couldn't load the satellite assembly which I think can be acceptable perf wise.

It is not just about perf. It is also about a number of corner cases that we need to deal with and that will not do what people want half of the time.

Comment (tarekgh) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: > Is there a default/standard casing for the culture name IDs

Yes but normalizing the name would tricky. We have to know if this culture is neutral culture or not to do the right casing ourselves. en-US is the casing for the specific cultures. while zh-Hant is for neutrals.  we can know if the culture is neutral or not in main stream cases but not when we are in Invariant mode. As I mentioned we can hard code the cases we know when we are invariant mode but we expect to add more as we encounter more. 

> Can we normalize to that in the tooling?
either the tooling or at runtime, we'll have to normalize it ourselves in some cases as I mentioned in the globalization mode.

> It is also about a number of corner cases that we need to deal with and that will not do what people want half of the time.

Is there examples of that? Just for my education.

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: > not when we are in Invariant mode

If you are in Invariant mode, you are saying that you do not care about globalization. This does not have to work well or at all when we are in invariant mode.

> Is there examples of that? 

For example, you can have conflicts that differ just by casing. How do you pick the winner?

Comment (tarekgh) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: > If you are in Invariant mode, you are saying that you do not care about globalization. This does not have to work well or at all when we are in invariant mode.

Unfortunately as we ship Alpine images which have invariant mode on by default, and these images got used in some scenarios while the users not aware about that and then they run into some problems like this resource loading. 

> For example, you can have conflicts that differ just by casing. How do you pick the winner?

We probe first for the exact casing. If we couldn't find the exact casing, then the first enumerated match would be reasonable anyway because I am not expecting people will create multiple resource assemblies with exact same name. 

To summarize, I think it is nice to support case insensitive culture names as a resource directories. If we have any complication with that, then I am ok not support it and clearly document the behavior.

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs (hunk 1 above)
Comment (sdmaclea) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: I had copied this code from `System.TimeZoneInfo.EnumerateFilesRecursively()`.  I didn't really look at changing it much because it was already tested and it worked.

Looks like you are right.

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs (hunk 2 above)
Comment (sdmaclea) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: I assumed the only purpose of the outer try block was to catch the case when `ArrayPool<byte>.Shared.Rent(bufferSize)` threw.

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs (hunk 5 above)
Comment (sdmaclea) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: We have to allocate anyway

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs (hunk 8 above)
Comment (sdmaclea) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: To try to minimize corner cases, when searching for case insensitive matches, I don't stop on the first match, but parse the whole directory. If there is more than one matching directory, I ignore them all.

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: This enumeration still does not feel right. We own the app publish format and can publish stuff in whatever format works best for runtime, e.g. we can lower-case the culture names and be done with it.

Comment (sdmaclea) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: > lower-case the culture names 

It would certainly make the culture lookup easier. It definitely is painful to do this sort of lookup.

Backwards compat?
.NET Standard compat?

Mayber @tarekgh's **"ok not support it and clearly document the behavior."** is better.

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: The satellite probing for default context in [BindSatelliteResourceByResourceRoots](https://github.com/dotnet/coreclr/blob/3b47496ef2656333684526cf71478253a207c1a5/src/binder/assemblybinder.cpp#L1248) does not seem to doing any case-insensitive probing either. Is there any case-insensitive probing for default context hidden somewhere else?

Comment (sdmaclea) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: I don't think so.  This was added in response to @tarekgh's feature request to support case insensitive lookup of culture names on all platforms. 

I still haven't tested, but I assume the Default context hits this code too to fire it `Resolving` event. Otherwise, I'll have to add that code too...

Comment (sdmaclea) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: > We own the app publish format and can publish stuff in whatever format works best for runtime, e.g. we can lower-case the culture names and be done with it.

If we did that, I guess we could support case exact match and lowercase match for back compat….

I am drafting the patch to remove case insensitive search.

Comment (sdmaclea) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs]: Remove case insensitive search and opened #24211

File: src/binder/clrprivbinderassemblyloadcontext.cpp [hunk 10]
```diff
@@ -65,9 +65,10 @@ HRESULT CLRPrivBinderAssemblyLoadContext::BindAssemblyByName(IAssemblyName     *
         //
         // 1) Lookup the assembly within the LoadContext itself. If assembly is found, use it.
         // 2) Invoke the LoadContext's Load method implementation. If assembly is found, use it.
-        // 3) Lookup the assembly within TPABinder. If assembly is found, use it.
-        // 4) Invoke the LoadContext's Resolving event. If assembly is found, use it.
-        // 5) Raise exception.
+        // 3) Lookup the assembly within TPABinder (except for ssatellite requests). If assembly is found, use it.
```
Comment (vitek-karas) [inline src/binder/clrprivbinderassemblyloadcontext.cpp]:
```suggestion
        // 3) Lookup the assembly within TPABinder (except for satellite requests). If assembly is found, use it.
```

File: src/binder/clrprivbinderassemblyloadcontext.cpp [hunk 11]
```diff
@@ -65,9 +65,10 @@ HRESULT CLRPrivBinderAssemblyLoadContext::BindAssemblyByName(IAssemblyName     *
...
+        // 4) Invoke the LoadContext's ResolveSatelliteAssembly method (for ssatellite requests). If assembly is found, use it.
```
Comment (vitek-karas) [inline src/binder/clrprivbinderassemblyloadcontext.cpp]:
```suggestion
        // 4) Invoke the LoadContext's ResolveSatelliteAssembly method (for satellite requests). If assembly is found, use it.
```

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs [hunk 12]
```diff
@@ -555,6 +555,37 @@ public void Dispose()
...
+
+            AssemblyName parentAssemblyName = new AssemblyName(assemblyName.Name);
+
+            Assembly? parentAssembly = LoadFromAssemblyName(parentAssemblyName);
+
+            if (parentAssembly == null)
+                return null;
+
+            // ResolveSatelliteAssembly should always be called on the ALC which loaded parentAssembly
+            Debug.Assert(this == GetLoadContext(parentAssembly));
```
Comment (vitek-karas) [inline src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: If this is true it should be in theory possible to pass the parent assembly from the caller... I assume that's just too complicated.

Will the above `LoadFromAssemblyName` invoke ALC code, or will we get it from a runtime cache early on?

Comment (sdmaclea) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: > pass the parent assembly from the caller … too complicated

It could be done. This seemed simpler. 

The managed caller is passing the parent assembly. I could rework the native interfaces to keep this information. 

> will we get it from a runtime cache

It should be returned from the runtime cache
We would get more benefit from optimizing the cache lookup path.

Comment (vitek-karas) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: If it's gonna come from the cache, then this is perfectly fine... please ignore :-)

**Comment by sdmaclea** [comment]
I added the lower case culture probe since there seems to be consensus that this may be a good long term approach.
This is passing my manual testing. I'll write formal corefx testing...

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs (hunk 12 above)
Comment (sdmaclea) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: > If this is true

This would be true for normal usage, but not always true.  In case an app tries to manually load a random satellite assembly w/o loading its parent, this assumption would be invalid.  I added back the original logic to get the ALC for the parent assembly to handle this rare case.

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs [hunk 13]
```diff
@@ -555,6 +555,45 @@ public void Dispose()
                 }
             }
         }
+
+        private Assembly? ResolveSatelliteAssembly(AssemblyName assemblyName)
+        {
+            // Called by native runtime when CultureName is not empty
+            Debug.Assert(assemblyName.CultureName?.Length > 0);
+
+            string cultureName = assemblyName.CultureName!;
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: This can be done later only once we need it.

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs [hunk 14]
```diff
@@ -555,6 +555,45 @@ public void Dispose()
...
+
+            string satelliteSuffix = ".resources";
+
+            if (assemblyName.Name == null || !assemblyName.Name.EndsWith(satelliteSuffix))
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: EndsWith is "current culture" by default. Pass in `StringComparison.InvariantCulture` ?

Comment (tarekgh) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: should it be OrdinalIgnoreCase even?

Comment (sdmaclea) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: `StringComparison.InvariantCulture` or `StringComparison.Ordinal` would seem interchangeable here. I guess I would choose `StringComparison.Ordinal` because it could be  faster.

`IgnoreCase`...  In normal code flow, `GetSatellite()` appends the lower case ".resources".  Tooling will create lower case ".resources". Ignoring case would allow customer to have unconventional filename if they manually tried to load the satellite. I am not sure we want to support that. It would be a poorly supported scenario.
